Development/delete and update options for schedule

CptHouston opened the pull request:
-- update an event
-- update events range
-- delete event occurrence and related events 


relevant updates are in
- SchedulesController.cs
- ScheduleService.cs

CptHouston committed:
Create dotnet-tools.json

CptHouston committed:
-- reworked postSchedule endpoint
-- currently only weekly pattern available

CptHouston committed:
Revert "Create dotnet-tools.json"

This reverts commit 34e5a519b4dfb4daab5f0549a28a34c558e8f05a.

CptHouston committed:
Delete profile.arm.json

CptHouston committed:
-- added methods for other patterns

CptHouston committed:
Merge branch 'development/scheduleRework' of https://github.com/CptHouston/WhatBackend into development/scheduleRework

CptHouston committed:
Conflict check

CptHouston committed:
Conflict

CptHouston committed:
Merge branch 'dev' into development/scheduleRework

CptHouston committed:
Updated ScheduledEventHandler

CptHouston committed:
Merge branch 'development/scheduleRework' of https://github.com/CptHouston/WhatBackend into development/scheduleRework

CptHouston committed:
UpdatedDB scripts

CptHouston committed:
-- bugFix

CptHouston committed:
-- updated parser to increase readability

CptHouston committed:
Added inharitance
-- updated bug with swagger

CptHouston committed:
CreateScheduleAsync refactoring
added subfolders + removed provate methods

CptHouston committed:
-- added filtering for scheduled events

CptHouston committed:
-- added filtering for scheduled events

CptHouston committed:
-- added filtering for scheduled events

CptHouston committed:
-- added get event occurrence by id to schedule controller

CptHouston committed:
-- added the list of related events to event occurence DTO

CptHouston committed:
-- added the list of related events to event occurence DTO

CptHouston committed:
-- added the list of related events to event occurence DTO

CptHouston committed:
-- bugfixed

CptHouston committed:
-- updated naming

CptHouston committed:
-- GetFullDataFromStorage set to public access

CptHouston committed:
--bugfix in StorageParser

CptHouston committed:
-- updated GetById method to include the list of scheduled events

CptHouston committed:
-- Added update and delete options

CptHouston committed:
-- Added update and delete options

CptHouston committed:
-- added validation

CptHouston committed:
-- updated algorithm for relativeMonthly

Hlib-mats committed:
merged get groups by date endpoints logic into get all student groups

Now you can indicate dates in GET /api/student_groups endpoint and all groups in this interval will be returned

filinskyi committed:
Logic for importing different entites moved in separate files

filinskyi committed:
Merge https://github.com/ita-social-projects/WhatBackend into Fix/ImportService

filinskyi committed:
Merge https://github.com/ita-social-projects/WhatBackend into Fix/ImportService

filinskyi committed:
Merge branch 'Fix/ImportService' of https://github.com/ita-social-projects/WhatBackend into Fix/ImportService

CptHouston committed:
-- updated accessibility for non-public eventhandler methods

filinskyi committed:
renamed all interfaces and classes Importers

filinskyi committed:
removed IXLSFileService as it no more needed

CptHouston committed:
-- updated scheduleEventHandler
(constructor now gets only pattern, rest is passed via params to GetEvents)
-- ScheduleEventFactory class created

CptHouston committed:
-- IScheduledEventHandlerFactory dependency inverted

CptHouston committed:
Merge branch 'dev' into development/scheduleRework

CptHouston committed:
-- bugfix

CptHouston committed:
-- naming updated

CptHouston committed:
-- Added description for PostSchedule endpoint

CptHouston committed:
-- updateEventOccurrence by ID created

CptHouston committed:
-- updateEventOccurrence by ID created

CptHouston committed:
-- branch rebased to /scheduleRework

CptHouston committed:
-- branch rebased to /filteringForEvents

CptHouston committed:
-- branch rebased to /filteringForEvents

CptHouston committed:
-- added update event occurrence endpoint
-- updated endpoint description

CptHouston committed:
-- added update event occurrence endpoint
-- updated endpoint description

CptHouston committed:
-- bugfix

CptHouston committed:
-- bugfix

CptHouston commented:
should be merged after https://github.com/ita-social-projects/WhatBackend/pull/358

CptHouston committed:
Merge pull request #337 from CptHouston/development/scheduleRework

Development/schedule rework

CptHouston committed:
-- added ScheduledEventId as a filtering option

CptHouston committed:
-- bugfix

CptHouston committed:
-- bugfix

filinskyi committed:
Merge branch 'dev' into Fix/ImportService

filinskyi committed:
Merge pull request #366 from ita-social-projects/Fix/ImportService

Fix/import service

Hlib-mats committed:
Merge pull request #365 from ita-social-projects/feature/MergeEndpointsGetGroupsByDateAndGetAllGroups

Merged get groups by date endpoints logic into get all student groups

Hlib-mats committed:
Added remark for endpoint

CptHouston committed:
Merge pull request #358 from CptHouston/development/filteringForEvents

Development/filtering for events

CptHouston committed:
-- syntax and code visual update + bugfix

CptHouston committed:
-- syntax and code visual update + bugfix

filinskyi commented on CharlieBackend.Api/Controllers/SchedulesController.cs:
```diff
@@ -37,74 +38,125 @@ public SchedulesController(IScheduleService scheduleService)
+        /// Updates a single event
         /// </summary>
+        /// <remarks>
+        /// Old instance os event occurance is replaced with the new one (id is the same). 
```
> Old instance **os** of event **occurance** is replaced with the new one (id is the same).

replace bold words:
os -> of 
occurance -> occurrence

CptHouston replied:
updated

filinskyi commented on CharlieBackend.Api/Controllers/SchedulesController.cs:
```diff
@@ -37,74 +38,125 @@ public SchedulesController(IScheduleService scheduleService)
         /// Deletes exact schedule
         /// </summary>
+        /// <remarks>
+        /// Removes cheduled events related to specified EventOccurrence, updates start and finish date accordingle
```
> Removes **cheduled** events related to specified EventOccurrence, updates start and finish date **accordingle**

replace bold words:
cheduled -> scheduled
accordingle -> accordingly

CptHouston replied:
updated

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs:
```diff
@@ -0,0 +1,84 @@
+
+        protected override DateTime GetStartDate(int index)
+        {
+            DateTime startDate = new DateTime(_source.EventStart.Year, _source.EventStart.Month, 7 * (int)_pattern.Index,
+                _source.EventStart.Hour, _source.EventStart.Minute, _source.EventStart.Second);
+
+            int offset = _pattern.DaysOfWeek[index] == DayOfWeek.Sunday ? 7 : (int)_pattern.DaysOfWeek[index];
+            int startDay = (int)startDate.DayOfWeek;
+
+            return startDate.AddDays(startDay <= offset
+                       ? -1 * (startDay + 7 - offset)
+                       : -1 * (startDay - offset));
```
magic number 7, may be change it to some constant or static field ,
as you doing here :
![image](https://user-images.githubusercontent.com/29517010/107416957-c51a7080-6b1d-11eb-8a49-4fd6076945dc.png)

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+            startDate = startDate.AddMonths(_pattern.Interval);
+
+            startDate = new DateTime(startDate.Year, startDate.Month, 7 * (int)_pattern.Index,
+                startDate.Hour, startDate.Minute, startDate.Second);
+
+            int offset = _pattern.DaysOfWeek[index] == DayOfWeek.Sunday ? 7 : (int)_pattern.DaysOfWeek[index];
+            int startDay = (int)startDate.DayOfWeek;
+
+            startDate = startDate.AddDays(startDay <= offset
+                       ? -1 * (startDay + 7 - offset)
+                       : -1 * (startDay - offset));
```
same suggestion as in upper comment

CptHouston replied:
updated

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerWeekly.cs:
```diff
@@ -0,0 +1,31 @@
+            
+            return result.AddDays(result.DayOfWeek <= _pattern.DaysOfWeek[index]
+                    ? _pattern.DaysOfWeek[index] - result.DayOfWeek
+                    : 7 - (int)result.DayOfWeek + (int)_pattern.DaysOfWeek[index]);
```
same suggestion as in upper comment

CptHouston replied:
updated

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,378 @@
+            if (request.Context.MentorID.HasValue && !await _unitOfWork.MentorRepository.IsEntityExistAsync(request.Context.MentorID.Value))
+            {
+                error = "Mentor does not exist";
+            }
+
+            if (request.Context.ThemeID.HasValue && !await _unitOfWork.ThemeRepository.IsEntityExistAsync(request.Context.ThemeID.Value))
+            {
+                error = "Theme does not exist";
+            }
+
+            switch (request.Pattern.Type)
+            {
+                case PatternType.Daily:
+                    break;
+                case PatternType.Weekly:
+                    if (request.Pattern.DaysOfWeek == null || request.Pattern.DaysOfWeek.Count == 0)
+                    {
+                        error = "Target days not provided";
+                    }
+                    break;
+                case PatternType.AbsoluteMonthly:
+                    if (request.Pattern.Dates == null || request.Pattern.Dates.Count == 0)
+                    {
+                        error = "Target dates not provided";
+                    }
+                    break;
+                case PatternType.RelativeMonthly:
+                    if ((request.Pattern.DaysOfWeek == null || request.Pattern.DaysOfWeek.Count == 0) 
+                        || (request.Pattern.Index != null ? request.Pattern.Index <= 0 : false))
+                    {
+                        error = "Target days not provided";
+                    }
+                    break;
+                default:
+                    error = "Pattern type not supported";
+                    break;
+            }
+
+            return error;
+        }
```
these checks will override previous messages.
as example if all of them are valid "error" message will contain only the last one message.

or return in each "if" "case"
or if you want to do all checks and accumulate all messages than use StringBuilder.

CptHouston replied:
accumulating errors is a good idea
updated

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,378 @@
+                return "requestDTO body must not be null";
+            }
+
+            string error = null;
+
+            if (request.StudentGroupId.HasValue && !await _unitOfWork.StudentGroupRepository.IsEntityExistAsync(request.StudentGroupId.Value))
+            {
+                error = $"No such theme id={request.StudentGroupId.Value}";
+            }
+
+            if (request.MentorId.HasValue && !await _unitOfWork.MentorRepository.IsEntityExistAsync(request.MentorId.Value))
+            {
+                error = $"No such mentor id={request.MentorId.Value}";
+            }
+
+            if (request.ThemeId.HasValue && !await _unitOfWork.ThemeRepository.IsEntityExistAsync(request.ThemeId.Value))
+            {
+                error = $"No such theme id={request.ThemeId.Value}";
+            }
+
+            if (request.EventEnd.HasValue && request.EventStart.HasValue && (request.EventEnd < request.EventStart))
+            {
+                error = $"StartDate must be less then FinisDate";
+            }
+
+            return error;
```
same case as previous comment

CptHouston replied:
updated

filinskyi commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,378 @@
+                return "Request must not be null";
+            }
+
+            string error = null;
+
+            if (request.CourseID.HasValue && !await _unitOfWork.CourseRepository.IsEntityExistAsync(request.CourseID.Value))
+            {
+                error = "Course does not exist";
+            }
+
+            if (request.GroupID.HasValue && !await _unitOfWork.StudentGroupRepository.IsEntityExistAsync(request.GroupID.Value))
+            {
+                error = "Group does not exist";
+            }
+
+            if (request.MentorID.HasValue && !await _unitOfWork.MentorRepository.IsEntityExistAsync(request.MentorID.Value))
+            {
+                error = "Mentor does not exist";
+            }
+
+            if (request.StudentAccountID.HasValue && !await _unitOfWork.StudentRepository.IsEntityExistAsync(request.StudentAccountID.Value))
+            {
+                error = "Student does not exist";
+            }
+
+            if (request.ThemeID.HasValue && !await _unitOfWork.ThemeRepository.IsEntityExistAsync(request.ThemeID.Value))
+            {
+                error = "Theme does not exist";
+            }
+
+            if (request.StartDate.HasValue && request.FinishDate.HasValue && (request.StartDate < request.FinishDate))
+            {
+                error = $"StartDate must be less then FinisDate";
+            }
+
+            return error;
+        }
```
same case as previous comment

CptHouston replied:
updated

Hlib-mats commented on CharlieBackend.Core/Entities/EventOccurrence.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace CharlieBackend.Core.Entities
+{
+    public partial class EventOccurrence : BaseEntity
```
why does this class partial?

CptHouston replied:
oversight from previous built
updated

Hlib-mats commented on CharlieBackend.Core/Entities/Theme.cs:
```diff
@@ -12,5 +12,7 @@ public Theme()
         public string Name { get; set; }
 
         public virtual ICollection<Lesson> Lessons { get; set; }
+
+        public virtual ICollection<ScheduledEvent> ScheduledEvents { get; set; }
```
Does theme actually need schedule?

CptHouston replied:
field added as a part of the foreign key constraint

CptHouston committed:
-- grammar mistakes
-- added const fields to base eventHandler
-- validation changed to use StringBuilder

CptHouston committed:
Merge branch 'development/DeleteAndUpdateOptionsForSchedule' of https://github.com/CptHouston/WhatBackend into development/DeleteAndUpdateOptionsForSchedule

CptHouston committed:
-- bugfix

CptHouston commented:
pr transferred https://github.com/ita-social-projects/WhatBackend/pull/402
